### PR TITLE
Manifest and preamble

### LINKED ISSUE #37: Batch Script preamble
# Description
Add the ability for users to input lines into the batch script generated by the ``Step._write_script``. Effectively, this is so users can do things like ``module load xyz`` and ``conda activate env``. 

# Justification
Users needing to load specific environments or modules prior to execution. 

# Implementation Strategy
Add the following to ``BatchSettings`` base class.
 - ``_preamble`` - list of str, created by calling ``add_preamble``
 - ``add_preamble(self, line)`` - adds string parameter line to ``self._preamble``

Then we need to add it to the ``_write_script`` method of Step. There are two ways I think this could be done. 

Create a subclass of Step called ``BatchStep`` and have all the batch step implementations inherit from ``BatchStep``.
 - ``QsubBatchStep``
 - ``SbatchStep``
 - ``CobaltBatchStep``
then implement a ``add_preamble`` method to the ``BatchStep`` class so it gets inherited from each. Should probably do some santization of user input. This would probably be the best long term

OR

implement it in each batch ``Step`` class individually. 

Additionally this might need to consider which shell the user is using.

### PR DESCRIPTION
**Pull request by al-rigazzi**
This PR addresses the proliferation of different entity types and corresponding behaviors.

* Implemented `Manifest` object which acts as a container of deployables with checks and filters.
* Added `preamble` to `BatchSettings` to allow users to add non-WLM lines to batch scripts

Closes #37

### CODE REVIEW

File: smartsim/control/controller.py
Comment (Spartee) [inline smartsim/control/controller.py:273]: I think since the manifest returns empty lists we can just use ``manifest.ensembles`` and ``manifest.models`` directly in the conditional.

File: smartsim/control/manifest.py
Comment (Spartee) [inline smartsim/control/manifest.py:58]: Should the user know about the manifest object? To me it seems like an internal representation. Would this error be confusing?

File: smartsim/control/manifest.py
Comment (Spartee) [inline smartsim/control/manifest.py:91]: When I refactor the jobManager with the different classes for jobs, we should put in a check where the manifest is passed to the JM to see if any new entities are defined by the same name

File: smartsim/generation/generator.py
Comment (Spartee) [inline smartsim/generation/generator.py:87]: Should we just pass the ``Manifest`` object itself?

**Review by Spartee — COMMENTED** [review]
Looks much cleaner. Lets make sure Coverage is still over 90%. Tests passed for me.

File: smartsim/control/manifest.py
Comment (al-rigazzi) [reply smartsim/control/manifest.py:58]: OK, maybe I can change the message to "Attempted to create two orchestrators"?

File: smartsim/control/manifest.py
Comment (al-rigazzi) [reply smartsim/control/manifest.py:91]: OK, it makes sense to me.

File: smartsim/generation/generator.py
Comment (al-rigazzi) [reply smartsim/generation/generator.py:87]: OK, I like the idea of unifying the interface.

File: smartsim/control/controller.py
Comment (al-rigazzi) [reply smartsim/control/controller.py:273]: Yes, we can do this. The only reason I did not was that currently, every time we call `manifest.ensembles` or `manifest.models`, the whole list of deployable is searched and filtered (i.e. the separate lists are not maintained, always recomputed). Therefore, doing the check and then iterating over the lists means computing them twice. Anyhow, given that I don't think the user will create lists of millions of deployables, this shouldn't impact performances.

Comment (al-rigazzi) [reply smartsim/control/controller.py:273]: NVM, we can totally skip the conditional, as we are iterating over the resulting list. If the list is empty, the iteration will simply not be run.

File: smartsim/generation/generator.py
Comment (al-rigazzi) [reply smartsim/generation/generator.py:87]: @Spartee I'm actually going to push back on this. `_gen_entity_dirs` is used in a slightly different way (because it also has an optional parameter) and passing a `Manifest` to it would generate duplicates.

Comment (Spartee) [reply smartsim/generation/generator.py:87]: Should we edit the generator code then to make the interface cleaner? My guess is we could make some of the code more re-usable? Maybe not in this ticket but file a future one? Im going to leave that up to you.

File: smartsim/settings/alpsSettings.py
Comment (Spartee) [inline smartsim/settings/alpsSettings.py:105]: thanks

**Review by Spartee — APPROVED** [review]
Minor comments, LGTM otherwise. good to go assuming coverage is good.

I did not re-run the tests this review since these were minor changes, fyi.

**Comment by al-rigazzi** [comment]
@Spartee thanks for the review. I'll add a flag for lsf-related code, to keep the coverage rate of the rest of the library untouched.

**Comment by al-rigazzi** [comment]
@Spartee ok, I'll check coverage, but it should be unaffected. Thanks for the review.

**Pull request merged**
